### PR TITLE
Indirect Diffraction - OSIRIS Diffspec calibration file option

### LINKED ISSUE #14794: OSIRIS diffspec doublet peaks
Vicky has reported that:
>I was contacting on OSIRIS last week and noticed something weird in the Diffspec option of the diffraction. It sums all detectors 3-962 by default, and I was getting some doublet peaks which made little sense. So, I clicked “Use individual groupings” and plotted some separately to find that the peak positions (and intensities) were different. I realise that this is probably due to the varying distances that the detectors are placed with respect to the sample (Franz confirmed this). It looks like this for example:
image

> If you use Diffonly and run the instrument in full diffraction mode then it’s fine because you have the Ariel calibration file, which I am assuming cleans this up.

### PR DESCRIPTION
**Pull request by ElliotAOram**
Add option to use a calibration file for OSIRIS diffraction DiffSpec mode has been added in the interface and the underlying algorithm.

**To test:**

Tests need to be performed for the algorithm and the interface. 
- for both tests: **Ensure you turn on data archive searching and you will also need to have the `osiris_041_RES10` is available from the unit tests folder.**

**Algorithm**
- The following script should test the algorithm.

```
Diffraction_ws = ISISIndirectDiffractionReduction(InputFiles='OSIRIS00118782.nxs', CalFile='osiris_041_RES10.cal', Instrument='OSIRIS', SpectraRange='3,962')
```
- The output will be a workspace group, plot its member and ensure it looks like the diagram below

**Interface**
- Change the default facility/instrument to ISIS/OSIRIS (View > Preferences > Mantid)
- Open Indirect Diffraction (Interfaces > Indirect > Diffraction)
- Change the instrument to OSIRIS 
- Change diffraction to `Diffspec` mode.
- Ensure the default number of spectra are `3` to `962` 
- Use the run number 118782
- Ensure the `Cal File` option is greyed out initially
- Check the `Use Calibration File` option  
- Use the `osiris_041_RES10` calibration file.
- Click `Run`
- Plot the output workspace and ensure the plot is the following
  image

Fixes #14794 

[RELEASE NOTES](https://github.com/mantidproject/mantid/blob/b119ee99a4a8b275d9bdf9bf0d7d230c8736ffe8/docs/source/release/v3.8.0/indirect_inelastic.rst)

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by samueljackson92** [comment]
I get a warning while building this on OSX:

```bash
[10/50] Building CXX object Framework/DataHandli...MakeFiles/DataHandling.dir/src/LoadSpice2D.cpp.o[43/50] Building CXX object MantidQt/CustomInter.../src/Indirect/IndirectDiffractionReduction.cpp.o
/Users/samueljackson/git/mantid-test/MantidQt/CustomInterfaces/src/Indirect/IndirectDiffractionReduction.cpp:691:9: warning: 'CHECK_VALID' macro redefined [-Wmacro-redefined]
#define CHECK_VALID(text, validator)                                           \
        ^
/Users/samueljackson/git/mantid-test/MantidQt/CustomInterfaces/src/Indirect/IndirectDiffractionReduction.cpp:629:9: note: previous definition is here
#define CHECK_VALID(text, validator)                                           \
        ^
1 warning generated.
```

**Comment by ElliotAOram** [comment]
Thanks @samueljackson92 I'll take a look at those and let you know when it's ready to test again

File: MantidQt/CustomInterfaces/src/Indirect/IndirectDiffractionReduction.cpp [hunk 1]
```diff
@@ -270,6 +285,16 @@ void IndirectDiffractionReduction::runGenericReduction(QString instName,
   // Set algorithm properties
   msgDiffReduction->setProperty("Instrument", instName.toStdString());
   msgDiffReduction->setProperty("Mode", mode.toStdString());
+
+  // Check if Cal file is used
+  if (instName == "OSIRIS") {
```
Comment (samueljackson92) [inline MantidQt/CustomInterfaces/src/Indirect/IndirectDiffractionReduction.cpp]: Why three nested if statements? Could they not be combined into one? Or collected into a bool with a descriptive name?

File: MantidQt/CustomInterfaces/src/Indirect/IndirectDiffractionReduction.cpp [hunk 2]
```diff
@@ -633,6 +663,48 @@ bool IndirectDiffractionReduction::validateVanCal() {
...
+* Checks to see if the cal file and optional rebin fields are valid.
+*
+* @returns True if calibration file and rebin values are valid, false otherwise
+*/
+bool IndirectDiffractionReduction::validateCalOnly() {
+  // Check Calib file valid
+  if (m_uiForm.ckUseCalib->isChecked()) {
+    if (!m_uiForm.rfCalFile_only->isValid())
+      return false;
+  }
```
Comment (samueljackson92) [inline MantidQt/CustomInterfaces/src/Indirect/IndirectDiffractionReduction.cpp]: Again, why two if statements?

File: MantidQt/CustomInterfaces/src/Indirect/IndirectDiffractionReduction.cpp [hunk 3]
```diff
@@ -633,6 +663,48 @@ bool IndirectDiffractionReduction::validateVanCal() {
...
+    if (rebinValid && rebStartTxt.toDouble() >= rebEndTxt.toDouble()) {
+      rebinValid = false;
+      m_uiForm.valRebinStart_CalibOnly->setText("*");
+      m_uiForm.valRebinEnd_CalibOnly->setText("*");
+    }
+  }
+
+  return rebinValid;
+
+  return true;
```
Comment (samueljackson92) [inline MantidQt/CustomInterfaces/src/Indirect/IndirectDiffractionReduction.cpp]: Extra return statement that is never reached?

**Comment by samueljackson92** [comment]
Tried this again this morning:
 - I no longer have a squished calibration input file box! Hurray!

But:
 - It still does not filter for .cal files
 - When I click the browse button for calibration files the "use calibration file" checkbox appears to become disabled? Come ask me to demo if you like.

**Comment by samueljackson92** [comment]
As we have both just observed the file extension search for whatever version of Qt we're currently using appears to be slightly buggy on OSX and the problem is not here. Everything else on this test works as expected and the code is good. :shipit:

**Pull request merged**